Multi song

theJenix opened the pull request:
Modified messenger system to read and write file transfers directly from/to disk, to avoid having file data in memory.

theJenix committed:
Modified messenger and to read/write files directly to and from disk.  This
avoids having to have files loaded into memory when transporting them across the
wire.

Updated tests and supporting files for this change.

Changed the bluetooth socket to insecure (FOR TESTING)

theJenix committed:
Merge branch 'master' of git://github.com/TheLastCrusade/fanClub into multi_song

Conflicts:
	SoundStream/src/com/lastcrusade/soundstream/audio/SingleFileAudioPlayer.java
	SoundStream/src/com/lastcrusade/soundstream/net/message/Messenger.java

theJenix committed:
ADded comments to SingleFileAudioPlayer regarding the use of file descriptor.

Cleaned up File transfer implementation in messenger, pulling out descriptive
methods and cleaning up dead code.

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/audio/SingleFileAudioPlayer.java:
```diff
@@ -95,7 +99,7 @@ private boolean isValidPath() {
             //This will fail and throw and Exception if the filepath is bad
             new File((new File(entry.getFilePath()).getParentFile().list())[0]).exists();
-            isValid = true;
+            isValid = true; 
```
whitespace after semicolon

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/audio/SingleFileAudioPlayer.java:
```diff
@@ -77,7 +78,10 @@ public void play() {
                 player.reset();
-                player.setDataSource(entry.getFilePath());
+                //changed to use the underlying file descriptor, because this doesnt want to work on a Samsung Galaxy S3
+                //..see http://stackoverflow.com/questions/1972027/android-playing-mp3-from-byte
```
change '..' to '  '

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/MessageThread.java:
```diff
@@ -63,8 +64,10 @@ public MessageThread(BluetoothSocket socket, Handler handler, String disconnectA
         
-        mmMessenger = new Messenger();
-        mmWriter    = new MessageThreadWriter(mmOutStream);
+//        this.messenger = new Messenger(context.getCacheDir()); // context being the Activity pointer
```
Remove this line?

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/message/Messenger.java:
```diff
@@ -44,7 +50,40 @@
+     * Used for the incoming buffer, to avoid frequent writes to the file
+     * 
+     */
+    private static final int MIN_READ_IN = 102400;
```
units?

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/message/Messenger.java:
```diff
@@ -44,7 +50,40 @@
+    private static final int MIN_READ_IN = 102400;
+    
+    /**
+     * Maximum size to read from a socket at a time.
```
Maximum size _in bytes_

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/message/Messenger.java:
```diff
@@ -44,7 +50,40 @@
+    private byte[] inBytes = new byte[MAX_READ_SIZE];
+
+    /**
+     * Maximum size to write to a socket at a time.
```
Maximum size _in bytes_

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/message/Messenger.java:
```diff
@@ -110,39 +155,182 @@ public void clearOutputBytes() {
+                if (isWaitingForNewMessage()) {
+                    this.messageLength = readAndConsumeLength();
+                    this.receivedMessage = null;
+//                    Log.i(TAG, "Receiving " + this.messageLength + " byte message");
```
Either log or remove

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/message/Messenger.java:
```diff
@@ -110,39 +155,182 @@ public void clearOutputBytes() {
+            if (!processed) {
+                readNext(input);
             }
+//            Log.d(TAG, "Residual buffer data: " + inputBuffer.size() + " bytes left in buffer");
```
log or remove

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/message/Messenger.java:
```diff
@@ -110,39 +155,182 @@ public void clearOutputBytes() {
+     * @throws IOException
+     */
+    private boolean readAndConsumeFile() throws IOException {
+        if (this.fileBytesLeft <= 0 && inputBuffer.size() >= SIZE_LEN) {
```
why are we progressing of there are few fileBytesLeft?

theJenix committed:
Cleaned up comments and dead code

Added ability to test if we can log in Messenger, so we can keep logging in
place while still retaining the ability to run our unit tests.

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/message/Messenger.java:
```diff
@@ -178,7 +385,9 @@ private boolean processAndConsumeMessage() throws InstantiationException, Illega
                 processed = true;
             } else {
                 //otherwise, it's a WTF
```
Did you want to remove these?

theJenix replied:
Nah, it's good now that it's covered with the conditional.

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/service/MessagingService.java:
```diff
@@ -268,16 +268,16 @@ public void handleMessage(int messageNo,
-                    FileOutputStream fos = new FileOutputStream(outputFile);
-                    fos.write(message.getSongData());
-                    fos.close();
+//                    File outputFile = createTempFile(message);
```
Remove these lines

theJenix commented:
Blanket comment: log messages in Messenger are now covered with a test to see if we can log.  This has to do with the fact that we use the Andorid junit runner, which removes the Android classes from the classpath.  This change should allow log messages to work on a device, while also allowing the unit tests to pass.

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/service/MusicLibraryService.java:
```diff
@@ -288,8 +288,7 @@ private void sendSongData(String fromAddr, long songId) {
             File   songFile   = new File(filePath);
-            byte[] bytes      = loadFile(songFile);
-            getMessagingService().sendTransferSongMessage(fromAddr, songId, songFile.getName(), bytes);
+            getMessagingService().sendTransferSongMessage(fromAddr, songId, songFile.getName(), songFile.getCanonicalPath());
```
Heads up this might conflict with @ejohnson44's changes

theJenix replied:
That's ok, my code is louder.

theJenix committed:
Killed dead code

theJenix committed:
Removed dead code

theJenix committed:
Pulled code at top of readAndConsumeFile into a helper method, for clarity

theJenix committed:
Merge branch 'master' of git://github.com/TheLastCrusade/fanClub into multi_song

Conflicts:
	SoundStream/src/com/lastcrusade/soundstream/audio/SingleFileAudioPlayer.java

Fixed compile error in PlaybarFragment due to changing getCurrentSong to
getCurrentEntry

reidbaker commented:
Could anyone else look at this tonight?

ejohnson44 commented on SoundStream/src/com/lastcrusade/soundstream/net/MessageThreadWriter.java:
```diff
@@ -57,39 +61,29 @@ public int compare(QueueEntry lhs, QueueEntry rhs) {
         qe.messageNo = messageNo;
-        qe.score     = messageNo * (TransferSongMessage.class.isAssignableFrom(messageType) ? 100 : 1);
-        qe.type      = messageType;
+        qe.score     = messageNo * (TransferSongMessage.class.isAssignableFrom(message.getClass()) ? 100 : 1);
```
Can you possible comment on the ternary things? I never quite remember how they work

theJenix replied:
You can read it as "if condition == true then use the first value else use the second value"  In this case, I want to multiply the messageNo by 100 for transfer song messages.  This prioritizes them far below other types of messages within the queue.

ejohnson44 commented on SoundStream/src/com/lastcrusade/soundstream/net/message/IFileMessage.java:
```diff
@@ -0,0 +1,21 @@
+package com.lastcrusade.soundstream.net.message;
```
Add in copyright

theJenix replied:
:finnadie:

ejohnson44 replied:
What on earth is that...?

theJenix replied:
One of the "Doom" faces...it's called finnadie...I'm guessing it means finished in some language.

ejohnson44 commented on SoundStream/src/com/lastcrusade/soundstream/net/message/Messenger.java:
```diff
@@ -217,4 +436,91 @@ private boolean processAndConsumeMessage() throws InstantiationException, Illega
+    }
+
+    public void writeToOutputStream(OutputStream outStream) throws IOException {
+        // TODO: this may or may not be needed...for now it does not appear it
```
can this be removed now?

theJenix replied:
The code below this todo may be required for the "break up transfer song message" action...I'd prefer to leave it in place for now.

theJenix committed:
added copyright to IFileMessage

reidbaker commented:
I am getting IO exceptions on the host because the socket it closing. Cant get through a normal walkthrough

theJenix committed:
Made readingFile a member variable, so we can save state beyond calls to
deserializeMessage

theJenix commented:
Unfortunately, I wasn't getting those errors on that configuration before the fix, but I could see how the error could occur.  After commit ad33614, this was tested successfully on 2013-04-02 using a Nexus 4 (guest) and Galaxy S3).  Should be good to :goberserk:

reidbaker commented:
I tested with a 2 Galaxy nexus's and looked good I am pulling it in
